File: Code Python/BetaProjet-master/Pattern_generator.py
import random as rd
from functions import rotation


class PatternGenerator:
    empty = [
        ['e', 'e', 'e'],
        ['e', 'e', 'e'],
        ['e', 'e', 'e']
    ]
    square = [
        ['w', 'w', 'e'],
        ['w', 'w', 'e'],
        ['e', 'e', 'e']
    ]
    corner_2 = [
        ['w', 'w', 'e'],
        ['e', 'e', 'e'],
        ['e', 'e', 'e']
    ]
    center_line = [
        ['e', 'w', 'e'],
        ['e', 'w', 'e'],
        ['e', 'w', 'e']
    ]
    center_point = [
        ['e', 'e', 'e'],
        ['e', 'w', 'e'],
        ['e', 'e', 'e']
    ]
    top_point = [
        ['e', 'w', 'e'],
        ['e', 'e', 'e'],
        ['e', 'e', 'e']
    ]
    corner_point = [
        ['w', 'e', 'e'],
        ['e', 'e', 'e'],
        ['e', 'e', 'e']
    ]
    side_line = [
        ['w', 'e', 'e'],
        ['w', 'e', 'e'],
        ['w', 'e', 'e']
    ]
    corner_3 = [
        ['w', 'w', 'e'],
        ['w', 'e', 'e'],
        ['e', 'e', 'e']
    ]
    anti_corner = [
        ['e', 'w', 'e'],
        ['w', 'w', 'e'],
        ['e', 'e', 'e']
    ]
    cross = [
        ['e', 'w', 'e'],
        ['w', 'w', 'w'],
        ['e', 'w', 'e']
    ]
    double_line = [
        ['w', 'w', 'w'],
        ['e', 'e', 'e'],
        ['w', 'w', 'w']
    ]
    double_side_point = [
        ['w', 'e', 'e'],
        ['e', 'e', 'e'],
        ['w', 'e', 'e']
    ]

    pattern_collection = dict()
    pattern_collection["empty"] = empty
    pattern_collection["square"] = square
    pattern_collection["corner_2"] = corner_2
    pattern_collection["center_line"] = center_line
    pattern_collection["side_line"] = side_line
    pattern_collection["center_point"] = center_point
    pattern_collection["top_point"] = top_point
    pattern_collection["corner_point"] = corner_point
    pattern_collection["side_line"] = side_line
    pattern_collection["corner_3"] = corner_3
    pattern_collection["anti_corner"] = anti_corner
    pattern_collection["cross"] = cross
    pattern_collection["double_line"] = double_line
    pattern_collection["double_side_point"] = double_side_point

    def __init__(self, patterns, weight=None):
        """
        contain a dict with every pattern selected an their associated weight
        :param patterns: list of patterns desired
        :param weight: list of weight for patterns selected
        """
        # PatternGenerator.pattern_collection = dict of every patterns with default weight of 1
        # self.pattern_selection = dict of selected patterns among the whole set
        self.pattern_selection = list(patterns)
        if weight is not None:
            assert len(weight) == len(patterns)
            for i in range(len(weight)):
                self.pattern_selection[i].append(weight[i])
        else:
            for i in range(len(patterns)):
                self.pattern_selection[i] = [self.pattern_selection[i]]
                self.pattern_selection[i].append(1)

    def assign_weight(self, key, weight):
        for el in self.pattern_selection:
            if el[0] == key:
                el[1] = weight

    def return_random(self):
        chosen_one = rd.choice(self.pattern_selection)[0]
        angle = rd.randint(0, 3) * 90
        return rotation(angle, self.pattern_collection[chosen_one])

    def return_semi_random(self):
        # random float between 0 and 1
        rnd = rd.random()
        # calculating the total sum of the weight then reducing it
        sum_weight = 0
        for el in self.pattern_selection:
            sum_weight += el[1]
        adjusted_sum = sum_weight * rnd

        # find the random pattern
        for item in self.pattern_selection:
            # item[0] = name / item[1] = weight
            adjusted_sum -= item[1]
            if adjusted_sum <= 0:
                angle = rd.randint(0, 3) * 90
                return rotation(angle, self.pattern_collection[item[0]])
        raise ValueError("Problem in function")

    def return_empty(self):
        return self.empty


if __name__ == "__main__":
    my_generator = PatternGenerator(["empty", "double_line", "center_point"])
    print("selection :\n", my_generator.pattern_selection)
    my_generator.assign_weight("empty", 0)
    print("\nassignation of weight:\n", my_generator.pattern_selection)
    print("\nchoice of a random el:\n", my_generator.return_random())
    print("\nchoice of a semi random el:\n", my_generator.return_semi_random())
    print("\nreturn of empty\n:", my_generator.return_empty())

File: src/main/java/be/ac/umons/bernardhofmanshouba/MapGeneration/PatternGenerator.java
package be.ac.umons.bernardhofmanshouba.MapGeneration;

import java.util.Random;

/*
patterns = new Pattern[] {
                Pattern.EMPTY,
                Pattern.CENTER_LINE,
                Pattern.SQUARE,
                Pattern.CORNER_2,
                Pattern.SIDE_LINE,
                Pattern.DOUBLE_LINE
        };
 */

/**
 * Class used to generate random pattern among a selection of pattern
 * each pattern has a probability attached to it
 */
public class PatternGenerator {
    private final Pattern[] patterns;
    private final double[] weights;

    public PatternGenerator(){
        patterns = new Pattern[] {
                Pattern.CENTER_LINE,
                Pattern.SQUARE,
                Pattern.SIDE_LINE,
                Pattern.CORNER_3,
                Pattern.EMPTY
        };
        weights = new double[]{
                1,
                0.2,
                1,
                0.2,
                0.1
        };
    }

    /**
     * @return a random pattern chosen among the selected
     */
    public char[][] getRandomPattern(){
        double rnd = (new Random()).nextDouble();

        double sumOfWeights = 0;
        for(double weight : weights){
            sumOfWeights += weight;
        }

        sumOfWeights *= rnd;

        for(int i = 0; i < patterns.length; i++) {
            sumOfWeights -= weights[i];
            if(sumOfWeights <= 0){
                return patterns[i].getPattern();
            }
        }
        throw new IllegalStateException("Not supposed to be here");
    }

    /**
     * Contains patterns that are used in the generation of the walls
     */
    private enum Pattern{
        EMPTY (new char[][]{
                {'e','e','e'},
                {'e','e','e'},
                {'e','e','e'}
        },PatternType.X),

        CENTER_POINT(new char[][]{
            {'e', 'e', 'e'},
            {'e', 'w', 'e'},
            {'e', 'e', 'e'}
        },PatternType.X),

        CROSS(new char[][]{
            {'e', 'w', 'e'},
            {'w', 'w', 'w'},
            {'e', 'w', 'e'}
        },PatternType.X),

        CENTER_LINE(new char[][]{
                {'e', 'w', 'e'},
                {'e', 'w', 'e'},
                {'e', 'w', 'e'},
        },PatternType.I),

        DOUBLE_LINE(new char[][]{
                {'w', 'w', 'w'},
                {'e', 'e', 'e'},
                {'w', 'w', 'w'}
        },PatternType.I),

        SQUARE(new char[][]{
                {'w', 'w', 'e'},
                {'w', 'w', 'e'},
                {'e', 'e', 'e'}
        },PatternType.E),

        CORNER_2(new char[][]{
                {'w', 'w', 'e'},
                {'e', 'e', 'e'},
                {'e', 'e', 'e'}
        },PatternType.E),

        TOP_POINT(new char[][]{
                {'e', 'w', 'e'},
                {'e', 'e', 'e'},
                {'e', 'e', 'e'}
        },PatternType.E),

        CORNER_POINT(new char[][]{
                {'w', 'e', 'e'},
                {'e', 'e', 'e'},
                {'e', 'e', 'e'}
        },PatternType.E),

        SIDE_LINE(new char[][]{
                {'w', 'e', 'e'},
                {'w', 'e', 'e'},
                {'w', 'e', 'e'}
        },PatternType.E),

        CORNER_3(new char[][]{
                {'w', 'w', 'e'},
                {'w', 'e', 'e'},
                {'e', 'e', 'e'}
        },PatternType.E),

        ANTI_CORNER(new char[][]{
                {'e', 'w', 'e'},
                {'w', 'w', 'e'},
                {'e', 'e', 'e'}
        },PatternType.E),

        DOUBLE_SIDE_POINT(new char[][]{
                {'w', 'e', 'e'},
                {'e', 'e', 'e'},
                {'w', 'e', 'e'}
        },PatternType.E);


        final char[][] pattern;
        final PatternType patternType;

        /**
         * A pattern contains a pattern and its type which indicates its symmetry
         * @param pattern char[][] array
         * @param type symmetry of the pattern
         */
        Pattern(char[][] pattern, PatternType type){
            this.pattern = pattern;
            this.patternType = type;
        }

        /**
         * @return return the pattern randomly rotated
         */
        char[][] getPattern() {
            final int i;
            switch (patternType){
                case E:
                    i = (new Random()).nextInt(4);
                    return rotation(pattern, i * 90);
                case I:
                    i = (new Random()).nextInt(2);
                    return rotation(pattern, i * 90);
                case X:
                    return pattern;
                default:
                    throw new IllegalStateException("Unexpected enum value");
            }
        }
    }

    /**
     * Used to indicate the symmetry of a pattern
     * the symmetry is the same as the letter used
     */
    private enum PatternType {
        X,
        I,
        E
    }

    /**
     * Rotates a pattern clockwise
     * @param pattern pattern to be rotated
     * @param angle angle of rotation
     * @return rotated pattern
     */
    protected static char[][] rotation(char[][] pattern, int angle){
        if(angle == 0){
            return pattern;
        }
        else{
            char[][] rotatedPattern = new char[3][3];
            for (int row =0; row < pattern.length; row ++){
                for (int col =0; col < 3; col ++){
                    rotatedPattern[row][col] = pattern[col][row];
                }
            }
            for (int row = 0; row < pattern.length; row++){
                char temp = rotatedPattern[row][0];
                rotatedPattern[row][0] = rotatedPattern[row][2];
                rotatedPattern[row][2] = temp;
            }
            return rotation(rotatedPattern, angle-90);
        }
    }

    /**
     * @return the empty pattern
     */
    protected static char[][] getEmpty(){
        return Pattern.EMPTY.getPattern();
    }
}
